Add macos interaction.pressKeys

mzgoddard opened the pull request:
- Add loadOsModule to load interaction and session modules depending on `process.platform`
- Define modules and command types in `lib/modules/types.js`
- Move existing interaction and session modules into `lib/modules/win32`
- Add `lib/modules/macos`
- Add `lib/helpers/macos` with types and functions to parse key combination into an applescript and execute it with osascript

jugglinmike commented on lib/helpers/load-os-module.js:
```diff
@@ -0,0 +1,25 @@
+ * @return {T}
+ * @template T
+ */
+function loadOsModule(name, deferredModules) {
+  const { platform } = process;
+  if (!(platform in deferredModules)) {
+    throw new Error(`No "${name}" module for platform "${platform}"`);
+  }
+
+  return deferredModules[platform]();
+}
```
I'm not convinced this amount of abstraction is justified. How about a thin wrapper around `require` that just prefixes with the value of `process.platform`--something we could use like `requireOSModule('interaction')`?

mzgoddard replied:
Isn't what you are suggesting more abstract than what I did?

-----

I see two issues with your suggestion:
- How do we pass `require`'s binding or execution context?
- We lose type checking on the non-static `require` call.

### On passing `require` execution context

How are you thinking `requireOSModule` would wrap `require`?

```js
const requireOSModule = osModuleRequire(require);
module.exports = requireOSModule('interaction');
```

```js
module.exports = requireOSModule(require, 'interaction');
```

```js
module.exports = require(osModule('interaction'));
```

```js
module.exports = require(osModule(__filename, 'interaction'));
```

### On type checking

A wrapper like you are suggesting isn't a static import statement. So its hard for type analysis to check it. #38 we can check the types of the modules defined this way. From what I know, we can't do that with any of my above suggestions.

jugglinmike replied:
> * How do we pass `require`'s binding or execution context?

Do we need to do this? Our current needs would be served by a `requireOSModule` that operates relative to its own `require` function, and under the file structure you're proposing, that would scale to use with additional AT Driver modules.

> * We lose type checking on the non-static `require` call.

How is this different from the implementation of `loadOSModule` proposed here? It seems like the type annotations in this patch (e.g. `/** @type {ATDriverModules.Interaction} */`) address the same problem and would work equally well for `requireOSModule`.

mzgoddard replied:
> > * How do we pass `require`'s binding or execution context?
> 
> Do we need to do this? Our current needs would be served by a `requireOSModule` that operates relative to its own `require` function, and under the file structure you're proposing, that would scale to use with additional AT Driver modules.

I don't think we need this, for it to execute, but the module would need to be `lib/modules`.

> > * We lose type checking on the non-static `require` call.
> 
> How is this different from the implementation of `loadOSModule` proposed here? It seems like the type annotations in this patch (e.g. `/** @type {ATDriverModules.Interaction} */`) address the same problem and would work equally well for `requireOSModule`.

Current `loadOSModule` can internally determine the type's of the modules. The proposed `requireOSModule` won't be able to do that. Typescript will compare the value returned by `loadOSModule` against the type specified in `@type` and error if the value cannot be converted. When it does that for the value from `requireOSModule`, it will convert from `any`, the universal type, and never produce a conversion error.

Stated another way, Typescript can determine the type of `require('./win32/interaction')` as `ATDriverModules.Interaction`, but not the type of ```require(`./${process.platform}/${moduleName}`)```, which will default to `any`.

We can omit the `@type` in `./lib/modules/interaction` if we want with `loadOSModule`.

Though for that to work I realize I need to update the uses of `@type` to the following:

```js
module.exports = /** @type {ATDriverModules.Session} */ (
  loadOsModule('interaction', {
```

Here's an example error with the above type mismatch.

```sh
> @bocoup/windows-sapi-tts-engine-for-automation@0.0.4 test-types
> tsc -p tsconfig.json

lib/modules/interaction.js:7:29 - error TS2352: Conversion of type 'Interaction' to type 'Session' may be a mistake because neither type sufficiently overlaps with the other. If this was intentional, convert the expression to 'unknown' first.
  Property '"session.new"' is missing in type 'Interaction' but required in type 'Session'.

7 module.exports = /** @type {ATDriverModules.Session} */ (
                              ~~~~~~~~~~~~~~~~~~~~~~~

  lib/modules/types.js:55:2
    55  *   "session.new": ATDriverModules.SessionNewSession,
        ~~~~~~~~~~~~~~~~~
    '"session.new"' is declared here.


Found 1 error in lib/modules/interaction.js:7
```

Here's what it looks like in my editor.

![Screenshot 2024-02-15 at 11 39 29 AM](https://github.com/w3c/aria-at-automation-driver/assets/833298/47b3d5e9-d0e9-4041-a6a5-0d5328add9c9)

jugglinmike replied:
Sounds like the type checking justifies this!

jugglinmike commented on lib/helpers/macos/applescript.js:
```diff
@@ -0,0 +1,117 @@
+  validateCommand(command);
+  const script = `with timeout of ${APPLESCRIPT_TIMEOUT} seconds
+    tell application "System Events"${command.keyCodes
+      .map(
+        code => `
+        key code ${KeyCode[code].toString()}${
+          command.modifiers.length > 1
+            ? ` using {${command.modifiers.map(mod => `${Modifier[mod]} down`).join(', ')}}`
+            : command.modifiers.length === 1
+              ? ` using ${Modifier[command.modifiers[0]]} down`
+              : ``
+        }`,
+      )
```
I don't think expressing this as a single Template Literal expression does much for readability. A few variable bindings with expressive names would help folks recognize the code that's being generated.

jugglinmike commented on lib/helpers/macos/applescript.js:
```diff
@@ -0,0 +1,117 @@
+      }
+    });
+
+    child.stdin.write(script.script);
```
This demonstrates an overloaded definition of the word "script," potentially confusing discussion amongst maintainers. While the immediate remedy would be to introduce a distinct term for one of the two concepts (e.g. `script` and `script.source`), I don't see why we need both in the first place. As long as there is only one "ScriptKind", what is the value in a container data structure? In other words, couldn't a script just be a string?

mzgoddard replied:
I think using a container is more maintainable. One thing we can do with the container that we can't do with a `string` is document the container. That type information can persist through different functions. Type analysis can give you references to what takes it as a parameter and what creates the type.

-----

I'll update the member to `source`. I really like that improvement. I was having a similar thought that calling it `script` was redundant.

jugglinmike replied:
Works for me

jugglinmike commented on lib/helpers/macos/applescript.js:
```diff
@@ -0,0 +1,117 @@
+ * @throws
+ * @template T
+ */
+const retryOnTimeout = (exports.retryOnTimeout = async function (action) {
+  for (let i = 0; i < APPLESCRIPT_RETRY; i++) {
+    try {
+      return await action();
+    } catch (error) {
+      if (!error || !error.message || !error.message.includes(APPLESCRIPT_TIMEOUT_ERROR)) {
+        throw error;
+      }
+    }
+  }
+});
```
If evaluation reaches the end of this function, it means the "action" could not be completed despite all the retry attempts. The function should throw in this case.

jugglinmike commented on test/helpers/macos/applescript.js:
```diff
@@ -0,0 +1,126 @@
+    assert.equal(await execScript(applescript('return "hello"')), 'hello\n');
+  });
+
+  test('retryOnTimeout', async () => {
```
We should have a test for the branch that concerns an unexpected error from the "action" function and a test for a true timeout.

mzgoddard replied:
By true timeout do you mean a timeout produced by `osascript` like through `execScript`? I'm not sure there is a way to reliably do this. I learned while trying to do this that Applescript's `with timeout` doesn't set a timeout for the duration of executing its body, but sets the timeout used by each statement inside the body from the default (of 120 seconds) to the new value.

For example I tried it with `delay 2` inside `with timeout of 1 seconds`. From what I can guess `delay` doesn't use the timeout (the default of the new one specified this way).

-----

I added a `await new Promise(resolve => setTimeout(resolve, 1)` to the original test.

jugglinmike replied:
Apologies; my request was a little vague. The test you've added named 'rethrows timeout after too many retries' is the test I was looking for, so we're good.

jugglinmike commented on test/helpers/macos/applescript.js:
```diff
@@ -0,0 +1,126 @@
+ * @param {T[]} tests
+ * @template T
+ */
+function testEach(name, nameDelta, testfn, tests) {
+  for (let i = 0; i < tests.length; i++) {
+    test(`${name}: ${nameDelta(tests[i], i)}`, async () => {
+      await testfn(tests[i], i);
+    });
+  }
+}
```
I'm not such a fan of the cognitive overheard of this kind of abstraction in test code. Even if it is less repetitive, it makes it harder to understand the intent. Tests written like the following example will help us focus on the application logic rather than the testing infrastructure:

```js
test('parseCodePoints: space', () => {
  assert.deepEqual(parseCodePoints([' ']), keyCodeCommand(['space'], []));
});
```

jugglinmike reviewed (changes requested):
[Years ago, in the context of the "harness" project, we agreed that unverifiable structured comments degrade the quality of a codebase.](https://github.com/w3c/aria-at-automation-harness/issues/3) Seeing the same patterns being introduced here in this project, my preference is to extend the automated validation process, but I'd settle for documentation that describes how contributors can perform the validation manually. Barring either of those solutions, we should remove the comments.

mzgoddard commented:
@jugglinmike, I made these changes:

- Rebased onto main and resolved conflict in `package.json`
- Added await to retryOnTimeout test
- Added retryOnTimeout test that forwards non timeout error
- Turned testEach into suites with individual `test('name', () => {` per testEach member in `test/helpers/macos/applescript.js`
- Throw in retryOnTimeout if it retries too many times
- Replace `Applescript.Script.script` with `.source`
- Separated renderScript template string into subfunctions and variables
- Fixed `@type` uses to get Typescript to type check

jugglinmike reviewed (approved):
Sorry for the delay, @mzgoddard; I got sidetracked by [a test failure which is also occurring on the `main` branch](https://github.com/w3c/aria-at-automation-driver/pull/41). Now that I know this patch is unrelated, we can merge it and take on that problem separately.

(By the way, the code style also doesn't validate on `main`--we really need to get CI set up!)
